jsdoc/tutorials/usingParameters.md
When to use parameters
Setting up parameters

jsdoc/tutorials/03_usingParameters.md
Parametric JSCAD designs unleash a new level of functionality, allowing you to build flexibility into 
your designs, so that they can easily adapt to different situations.

## ParametricBox.js
```javascript
const jscad = require('@jscad/modeling')
const { booleans, primitives } = jscad

const getParameterDefinitions = () => {
  return [
    {name: 'outerWidth', caption: 'Outer width of box:', type: 'float', initial: 120},
    {name: 'outerDepth', caption: 'Outer depth of box:', type: 'float', initial: 100},
    {name: 'outerHeight', caption: 'Outer height of box:', type: 'float', initial: 50},
    {name: 'wallThickness', caption: 'Wall Thickness:', type: 'float', initial: 2},
    {name: 'cornerRadius', caption: 'Inside Corner Radius:', type: 'float', initial: 5},
   ];
}

const main = (params) => {
	return booleans.subtract(
		outerBox(params),
		innerBox(params)
	);
}

const outerBox = (params) => {
	return primitives.cuboid({
		size: [params.outerWidth, params.outerDepth, params.outerHeight],
		center: [0, 0, params.outerHeight / 2]
	});
}

const innerBox = (params) => {
	const size = [
		params.outerWidth - (2 * params.wallThickness),
		params.outerDepth - (2 * params.wallThickness),
		params.outerHeight + (2 * params.cornerRadius)
	];
	const center = [0, 0, (size[2] / 2) + params.wallThickness];

	return primitives.roundedCuboid({ size, center, roundRadius: params.cornerRadius, segments: 32 });
}

module.exports = { main, getParameterDefinitions }
```
<img src="img/parametricBox.png" alt="Parametric Box example">

## All Parameter Types:
Below is a comprehensive example of all the types of parameters you can specify in JSCAD:
```javascript
const getParameterDefinitions = () => {
  return [
    { name: 'group1', type: 'group', caption: 'Group 1: Text Entry' },
    { name: 'text', type: 'text', initial: '', size: 20, maxLength: 20, caption: 'Plain Text:', placeholder: '20 characters' },
    { name: 'int', type: 'int', initial: 20, min: 1, max: 100, step: 1, caption: 'Integer:' },
    { name: 'number', type: 'number', initial: 2.0, min: 1.0, max: 10.0, step: 0.1, caption: 'Number:' },
    { name: 'date', type: 'date', initial: '2020-01-01', min: '2020-01-01', max: '2030-12-31', caption: 'Date:', placeholder: 'YYYY-MM-DD' },
    { name: 'email', type: 'email', initial: 'me@example.com', caption: 'Email:' },
    { name: 'url', type: 'url', initial: 'www.example.com', size: 40, maxLength: 40, caption: 'Url:', placeholder: '40 characters' },
    { name: 'password', type: 'password', initial: '', caption: 'Password:' },

    { name: 'group2', type: 'group', caption: 'Group 2: Interactive Controls' },
    { name: 'checkbox', type: 'checkbox', checked: true, initial: '20', caption: 'Checkbox:' },
    { name: 'color', type: 'color', initial: '#FFB431', caption: 'Color:' },
    { name: 'slider', type: 'slider', initial: 3, min: 1, max: 10, step: 1, caption: 'Slider:' },
    { name: 'choice1', type: 'choice', caption: 'Dropdown Menu:', values: [0, 1, 2, 3], captions: ['No', 'Yes', 'Maybe', 'So so'], initial: 2 },
    { name: 'choice3', type: 'choice', caption: 'Dropdown Menu:', values: ['No', 'Yes', 'Maybe', 'So so'], initial: 'No' },
    { name: 'choice2', type: 'radio', caption: 'Radio Buttons:', values, captions, initial: 5 },

    { name: 'group3', type: 'group', initial: 'closed', caption: 'Group 3: Initially Closed Group' },
    { name: 'checkbox2', type: 'checkbox', checked: true, initial: '20', caption: 'Optional Checkbox:' }
  ]
}
```

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

## 0.6.0 (2017-05-xx)
total rework of application and library structure:
- modular, smaller codebase (node.js modules) instead of monolithic files
- actual npm package, dependency managment now done correctly via package.json
- shared browser/Node.js code base
- split out CSG.js scad-api, formats handling etc to seperate repositories
- moved all repositories to the [jscad organization](https://github.com/jscad)
- default stl output is now in binary format (more efficient)
- lots more internal changes and improvements

## 0.5.2 (2017-10-01)
updated documentation links by Z3 Dev, updated Ace editor to 1.2.4, fixed AMF export to set colors only when provided, enhanced Processor constructor to support Viewer options, added big.html to provide an example of using Processor and Viewer options, enhanced Processor to retain multiple returned objects, fixed difference() and intersection() functions for CAG by fischman

## 0.5.1 (2016-06-27)
refactored AMF import and export, enhanced STL import by adding support for MM colors by Z3 Dev,added local storage by Robert Starkey

## 0.5.0 (2016-05-01)
 added SVG import and export, added options to Processor and View classes, allow more flexibility in HTML by Z3 Dev

## 0.4.0 (2016-02-25)
refactored, functionality split up into more files, mostly done by Z3 Dev

## 0.3.1 (2015-10-23)
including new parameter options by Z3 Dev

## 0.3.0 (2015-07-02)
  format.js (Stefan Baumann), and Blob.js/openjscad improved by Z3 Dev

## 0.2.4 (2015-05-20)
  renumbering, latest csg.js from http://joostn.github.com/OpenJsCad/ adapted

## 0.024 (2015-04-08)
  dev branch opened

## 0.023 (2015-02-14)
  bumping version based on openscad.js

## 0.020 (2015-02-04)
  browser window resizing done properly, thanks to Z3 devs via pull request

## 0.019 (2015-01-07)
  various pull requests from github merged again

## 0.018 (2014-10-05)
  various pull requests from github merged

## 0.017 (2013-04-11)
  alpha channel supported in color() and .setColor()

## 0.016 (2013-04-07)
  csg.js: solidFromSlices() and .setColor() on polygon level, and examples by Eduard Bespalov

## 0.015 (2013-04-05)
  rudimentary AMF export and import, web and cli

## 0.014 (2013-04-03)
  multiple files via drag & drop, developing locally

## 0.013 (2013-04-01)
  include() on web-online & drag & drop (but not off-line) and cli (server-side)

## 0.012 (2013-03-20)
  improved UI (slider from the left)

## 0.011 (2013-03-28)
  added support for rectangular_extrude(), rotate_extrude() and torus()

## 0.010 (2013-03-22)
  leave .scad file intact, and translate on-the-fly

## 0.009 (2013-03-20)
  OpenSCAD .scad syntax support included via [openscad-openjscad-translator](https://github.com/garyhodgson/openscad-openjscad-translator) module, on web and cli; and experimental .stl import support (binary & ascii)

## 0.008 (2013-03-15)
  circle(), square(), polygon() partially and linear_extrude() implemented (openscad-like)

## 0.007 (2013-03-14)
  integrating jQuery for new features; draggable hint window

## 0.006 (2013-03-12)
  included examples available in the web-frontend direct

## 0.005 (2013-03-12)
  supporting webgui parameters as of original OpenJsCad (see examples/example030.jscad)

## 0.004 (2013-03-11)
  openscad.js: many improvements, more OpenSCAD-like functions

## 0.003 (2013-03-10)
  solidify the functionality (few bug fixes)

## 0.001 (2013-03-10)
  initial version


src/cli/cli.js
#!/usr/bin/env node
//--log_all

// NOTE: this will only run on Node > 6 or needs to be transpiled, or launched via launch-cli

// == OpenJSCAD.org CLI interface, written by Rene K. Mueller <spiritdude@gmail.com>, Licensed under MIT License
//
// Description:
//   openjscad <file> [-of <format>] [-o <output>]
// e.g.
//   openjscad test.jscad
//   openjscad test.jscad -o test.stl
//   openjscad test.jscad -o test.amf
//   openjscad test.jscad -o test.dxf
//   openjscad test.scad -o testFromSCAD.jscad
//   openjscad test.scad -o test.stl
//   openjscad test.stl -o test2.stl      # reprocessed: stl -> jscad -> stl
//   openjscad test.amf -o test2.jscad
//   openjscad test.jscad -of amf
//   openjscad test.jscad -of dxf
//   openjscad test.jscad -of stl
//   openjscad name_plate.jscad --name "Just Me" --title "CEO" -o amf test.amf
//
// History:
// 2016/10/01: 0.5.2: changes for libraries
// 2016/06/27: 0.5.1: refactored AMF import and export by Z3 Dev
//                    enhanced STL import, adding support for MM colors by Z3 Dev
// 2016/02/02: 0.4.0: GUI refactored, functionality split up into more files, mostly done by Z3 Dev
// 2015/07/02: 0.3.0: node 0.10.x support, /usr/bin/nodejs, new files involved: formats.js (Stefan Baumann) and Blob.js (Z3 Dev)
// 2014/12/09: 0.019: support of DXF output for 2D objects (laser cutter)
// 2013/04/25: 0.010: support of params passed to main()
// 2013/04/12: 0.008: reimplement parseAMF without jquery
// 2013/04/11: 0.007: support of alpha for AMF addded, bumping version
// 2013/04/05: 0.006: support of AMF added, requires node 0.8.1+
// 2013/03/25: 0.005: more sanity check on input and local installation support
// 2013/03/18: 0.004: STL .stl (binary & ascii) support (experimental via openscad.js)
// 2013/03/18: 0.003: OpenSCAD .scad support by Gary Hodgson's openscad-openjscad-translator module
// 2013/03/02: 0.002: proper installation of the dependencies (csg.js & openscad.js) so openjscad can be used properly
// 2013/03/01: 0.001: initial version, with base function from openscad.jscad
//
const fs = require('fs')
const {version} = require('../jscad/version')
const env = require('./env')
const generateOutputData = require('./generateOutputData')
const { formats } = require('../io/formats')

/*
const getExtension = filename => {
  const ext = extname(filename || '').split('.')
  return ext[ext.length - 1]
} */

// var csg = sphere(1);          // -- basic test
// var csg = require(file).main; // -- treating .jscad as module? later perhaps

const args = process.argv.splice(2)

// handle arguments
// inputs, outputs
let {inputFile, inputFormat, outputFile, outputFormat, params} = parseArgs(args)

// outputs
const output = determineOutputNameAndFormat(outputFormat, outputFile)
outputFormat = output.outputFormat
outputFile = output.outputFile

console.log(`converting ${inputFile} -> ${outputFile} (${formats[outputFormat].description})`)

let src = fs.readFileSync(inputFile, inputFile.match(/\.stl$/i) ? 'binary' : 'UTF8')

// -- convert from JSCAD script into the desired output format
// const outputData = generateOutputData(src, params, {outputFormat})
// -- and write it to disk
// writeOutputDataToFile(outputFile, outputData)
generateOutputData(src, params, {outputFile, outputFormat, inputFile, inputFormat, version})
  .then(function (outputData) {
    writeOutputDataToFile(outputFile, outputData)
  }).catch(error => console.error(error))

// -- helper functions ---------------------------------------------------------------------------------------
function parseArgs (args) {
  // hint: https://github.com/substack/node-optimist
  //       https://github.com/visionmedia/commander.js
  //
  // process.argv.forEach(function (val, index, array) {
  //  console.log(index + ': ' + val)
  // })
  if (args.length < 1) {
    console.log('USAGE:\n\nopenjscad [-v] <file> [-of <format>] [-o <output>]')
    console.log('\t<file>  :\tinput file (Supported types: .jscad, .js, .scad, .stl, .amf, .obj, .gcode, .svg, .json)')
    console.log('\t<output>:\toutput file (Supported types: .jscad, .stl, .amf, .dxf, .svg, .json)')
    console.log("\t<format>:\t'jscad', 'stla' (STL ASCII, default), 'stlb' (STL Binary), 'amf', 'dxf', 'svg', 'json'")
    process.exit(1)
  }

  let inputFile
  let inputFormat
  let outputFile
  let outputFormat
  let params = {}

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '-of') { // -of <format>
      outputFormat = args[++i]
    } else if (args[i].match(/^-o(\S.+)/)) { // -o<output>
      outputFile = args[i]
      outputFile = outputFile.replace(/^\-o(\S+)$/, '$1')
    } else if (args[i] === '-o') { // -o <output>
      outputFile = args[++i]
    } else if (args[i].match(/^--(\w+)=(.*)/)) { // params for main()
      params[RegExp.$1] = RegExp.$2
    } else if (args[i].match(/^--(\w+)$/)) { // params for main()
      params[RegExp.$1] = args[++i]
    } else if (args[i].match(/^--(\w+)$/)) { // params for main()
      params[RegExp.$1] = args[++i]
    } else if (args[i].match(/.+\.(jscad|js|scad|stl|amf|obj|gcode|svg|json)$/i)) {
      inputFile = args[i]
      inputFormat = RegExp.$1
      if (!fs.statSync(inputFile).isFile()) {
        console.log('ERROR: cannot open file <' + inputFile + '>')
        process.exit(1)
      }
    } else if (args[i].match(/^-v$/)) { // show the version and the environment information
      env()
      console.log('OpenSCAD Compatibility (' + version + ')')
    } else {
      console.log('ERROR: invalid file name or argument <' + args[i] + '>')
      console.log("Type 'openjscad' for help")
      process.exit(1)
    }
  }
  // exit if a input file was not provided
  if (inputFile === null) process.exit(1)

  if (!outputFormat && !outputFile) {
    outputFormat = 'stla'
  }

  return {
    inputFile,
    inputFormat,
    outputFile,
    outputFormat,
    params}
}

function determineOutputNameAndFormat (outputFormat, outputFile) {
  if (!outputFormat && outputFile && outputFile.length && outputFile.match(/\.(jscad|js|stl|amf|dxf|svg)$/)) { // output filename set
    outputFormat = RegExp.$1
  } else if (!outputFormat && outputFile && outputFile.length) { // output filename isn't valid
    console.log('ERROR: invalid output file <' + outputFile + '>')
    process.exit(1)
  } else if (outputFormat.match(/(jscad|js|stl|stla|stlb|amf|dxf|svg)/i)) { // output format defined?
    let ext = RegExp.$1
    if (!outputFile) { // unless output filename not set, compose it
      ext = ext.replace(/stl[ab]/, 'stl') // drop [ab] from stl
      outputFile = inputFile
      outputFile = outputFile.replace(/\.([^\.]+)$/, '.' + ext) // compose output filename
    }
  } else {
    console.log(`ERROR: invalid output format <${outputFormat}`)
    process.exit(1)
  }
  return {outputFormat, outputFile}
}

function writeOutputDataToFile (outputFile, outputData) {
  fs.writeFile(outputFile, outputData.asBuffer(),
    function (err) {
      if (err) {
        console.log('err', err)
      } else {
        // console.log('success')
      }
    }
  )
}
